### ISSUE TITLE
Insteon: Add Support for Leak Sensor

### ISSUE THREAD

**Issue by krkeegan** [hollie/misterhouse#437]
I know some users are already using leak sensors, I imagine you have the programmed as simple triggerlincs or something similar.

In my email inbox, I have some notes about the leak sensors being finicky.  My recollection is that they have a wet and dry group and the sensors only sends on commands to each group, but never off commands.  My notes suggest that the sensor doesn't always send these messages accurately??  Did anyone ever get this resolved?  Are people still manually changing the state of the leak sensors inside MH?

**Comment by GregGTH** [comment]
Hi,
I'm considering a water leak sensor for my dishwasher. Is there any progress on implementing a mh object that makes it easy to work with this sensor?
Thanks,
Greg

**Comment by peloy** [comment]
Hi Greg,

On Wed, Feb 18, 2015 at 11:55:19PM -0800, GregGTH wrote:

> Hi,
> I'm considering a water leak sensor for my dishwasher. Is there any
> progress on implementing a mh object that makes it easy to work with
> this sensor?

I am using the INSTEON leak sensor like this:

items.mht:

# Dishwasher
INSTEON_TRIGGERLINC, 32.c2.b3:01, dishwasher_leak_dry, Leak_Sensors # 2213 v2.0
INSTEON_TRIGGERLINC, 32.c2.b3:02, dishwasher_leak_wet, Leak_Sensors # 2213 v2.0
INSTEON_TRIGGERLINC, 32.c2.b3:04, dishwasher_leak_heartbeat, Leak_Sensors # 2213 v2.0

User code:

if ($state = state_now $dishwasher_leak_dry) {
    print_log "dishwasher_leak_dry: $state";
    net_mail_send to => 'xyz@example.com', subject => "dishwasher_leak_dry: $state<EOM>";
}

if ($state = state_now $dishwasher_leak_wet) {
    print_log "dishwasher_leak_wet: $state";
    net_mail_send to => 'xyz@example.com', subject => "dishwasher_leak_wet: $state<EOM>";
}

It's not ideal and far from perfect but it's a start. The other day
I got a flood of email messages because the code will send an email
every time there is a state change and for some reason it seems like the
sensor kept sending messages when it got wet.

I am not currently doing anything with the heartbeat messages. I see
them in the log file (once a day) and I should generate an alert (like
an email) when I do not receive a heartbeat message in one day. I'll
code that up at some point.

To summarize, there currently is no "INSTEON_LEAKSENSOR" device that you
can put in your items.mht file. I am not even sure it's necessary given
that the INSTEON_TRIGGERLINC device seems to get the job done, though.

Cheers,

Eloy Paris.-

**Comment by GregGTH** [comment]
Thanks, Eloy. I think that will work well enough!

**Comment by waynieack** [comment]
I know this thread is really old but I wrote a sub that I think will work for all the leak sensors. What I've seen with my testing was that the ones I have never send a dry (group 1) on unless you push the button.  When mine detects water, it sends 1 wet (group 2) on, then when it drys, it starts sending a wet on every 10 seconds. So I think that depending on the version, they do something a little different. 

With the code below, if you want an alert for missing heartbeats, in the MH web, enable developer mode with the gear in the upper right, go to the "m_dishwasher_sensor_hb" object, click on it and click the + next to "Schedule Control" and set the time and days you want to check for a heartbeat for that day (1 per day) and select the "on". This way you will only get an alert for the time you selected instead of 25 hours from the last heartbeat which could be at any time. I have my scheduled for sun-sat at 5pm which is a good time for me to get an alert if a sensor is not responding.

I use the m_dishwasher_sensor_alerting object so I have something in the MH web to easily disable the water/dry alerts with. There is nothing worse than having something go wrong and kill my phone with alerts that I can't stop easily. 

I use Pushover "$push->notify" for alerting, but you can change that to your method of alerting (email, etc).   


**In my .mht for each sensor:**



```
GROUP,          Leak_Sensors
INSTEON_TRIGGERLINC, 40.xx.xx:01, dishwasher_sensor_dry, Leak_Sensors # 2852 v2.6
INSTEON_TRIGGERLINC, 40.xx.xx:02, dishwasher_sensor_wet, Leak_Sensors # 2852 v2.6
INSTEON_TRIGGERLINC, 40.xx.xx:04, dishwasher_sensor_heartbeat, Leak_Sensors # 2852 v2.6
```



**In my user code for each leak sensor:**


```
#noloop=start
$Leak_Sensor_Monitor = new Group(); # group for monitor objects (only needed 1 time for all sensors)
$m_dishwasher_sensor = new Generic_Item; # object used to track the state without getting duplicate state changes IE: the group 1 on every 10 sec.
$m_dishwasher_sensor_hb = new Generic_Item; # Object used to schedule daily HB checks/alerts in the MH web
$m_dishwasher_sensor_alerting = new Generic_Item; # Object used to disable alerting from the MH web
$Leak_Sensor_Monitor->add($m_dishwasher_sensor);
$Leak_Sensor_Monitor->add($m_dishwasher_sensor_hb);
$Leak_Sensor_Monitor->add($m_dishwasher_sensor_alerting);
#noloop=stop


&leaksensor('dish washer',$m_dishwasher_sensor,$m_dishwasher_sensor_hb,$m_dishwasher_sensor_alerting,$dishwasher_sensor_wet,$dishwasher_sensor_dry,$dishwasher_sensor_heartbeat);
```



**Add this sub to your user code and it can be used for all of your sensors:**
```
sub leaksensor {
 #<long name for alerting message>,<Generic object for status tracking>,<Generic object for heartbeat tracking/alert schedule>,
 #<Generic object to disable/enable alerting in MH web>,<Insteon wet object/group 2>,<Insteon dry object/group 1>,<Insteon heartbeat object/group 4>

  my ($longname,$monitor_object,$monitor_object_hb,$monitor_object_alerting,$wet_object,$dry_object,$hb_object) = @_;

  if ($Startup or $Reload) {
        $monitor_object_hb -> set_states('on','off');
        $monitor_object_alerting -> set_states('on','off');
  }

  if ( state_now $wet_object eq 'on' ) {
     print_log "$longname water sensor sent wet on";
     set $monitor_object 'wet' unless state $monitor_object eq 'wet';
  } elsif ( state_now $dry_object eq 'on' ) {
     print_log "$longname water sensor sent dry on";
     set $monitor_object 'dry' unless state $monitor_object eq 'dry';          # It only seems to send a dry on when the button is pushed when its dry.
  }

  if ( state_now $monitor_object_hb eq 'on' ) {
      $monitor_object_hb->set('off');
      if ( $hb_object->time_idle('25 hours') ) {
            print_log "$longname water sensor sent no heartbeats in the last 25 hours";
            $push->notify( "$longname water sensor sent no heartbeats in the last 25 hours", { title => 'Moisture Sensor HB Failure', priority => 1 });
         }
  }

  if ( ( new_minute(5) ) and ( $wet_object->time_idle('4 minutes') ) and ( $dry_object->time_idle('4 minutes') )  ) {
      set $monitor_object 'off' unless ( state $monitor_object eq 'off' ); # Reset monitor object when no messages from the sensor, this is a fail safe
  }


  if ( state_now $monitor_object eq 'wet' ) {
      print_log "$longname water sensor detected water!!";
      $push->notify( "$longname water sensor detected water!!", { title => 'Water Detected!!', priority => 2 }) unless ( state $monitor_object_alerting eq 'off' );
  } elsif ( state_now $monitor_object eq 'dry' ) {
      print_log "$longname water sensor is now DRY.";
      $push->notify( "$longname water sensor is now DRY.", { title => 'Water Sensor Dry', priority => 2 }) unless ( state $monitor_object_alerting eq 'off' );
  }

}
```

**Comment by GregGTH** [comment]
Hi Wayne,
I was just going through old lists and saw your reply! I never did get my
leak sensor to do much of anything. I concluded that it was faulty, but
never pursued it further. I've since moved, but the next time I have a need
for a leak sensor, I will try out this great-sounding module!

-Greg

**Comment by waynieack** [comment]
The leak sensor is for sure very odd, but I've been using my code for a
while now ( even got a real leak alert from my hot water heater )  and it
seems to work very well. The only thing that does not work as I would like
is when the sensor gets wet, you get alerted but until you push the button
on the device, you will get a heartbeat failure every 24 hours because the
sensor sends the heartbeats to a different group than when its dry. I guess
its not that a big deal because it reminds you to go push the button to
clear the wet alarm, but if you don't know why its happening it seems like
something is wrong with it. I decided not to try and fix the heartbeat
issue described above because it requires some guess work and I would
rather get more alerts than miss something when it comes to a leak.

_Wayne
